modules/standard/Version.chpl

```diff
@@ -0,0 +1,349 @@
+    official Chapel release, this will have type ``semanticVersion``
+    while for a pre-release, it will have type ``sourceVersion``.
+
+    Note that Chapel ``1.x.y``/``2.x.y`` corresponds to version
+    ``0.x.y``/``1.x.y`` in traditional semantic versioning.
+  */
+
+  const chplVersion;
+  chplVersion = version(chplMajor, chplMinor, chplUpdate, chplSHA);
+
+
+  /*
+    A helper function that creates a new version value from its
+    arguments.
+
+    :arg major: The major version number
+    :type major: `int`
+
+    :arg minor: The minor version number
+    :type minor: `int`
+
+    :arg update: The optional update version number (defaults to 0)
+    :type update: `int`
+
+    :arg commit: The optional commit ID (defaults to "")
+    :type commit: `string`
+
+    :returns: A new version value.  If commit == "", this routine
+              returns a value of :type:`semanticVersion` otherwise it
+              returns a value of :type:`sourceVersion`.
+  */
+  proc version(param major: int, param minor: int,
+               param update: int = 0, param commit: string = "") {
+    if commit == "" then
+      return new semanticVersion(major, minor, update);
+    else
+      return new sourceVersion(major, minor, update, commit);
+
+  }
+
+
+  /*
+    This record represents a semantic version value.  It uses
+    ``param`` values to represent its components in order to support
+    compile-time comparison of version numbers which in turn permits
+    code to specialize to specific versions of Chapel.  When printed
+    or converted to a string, it is represented as ``version
+    major.minor.update``.
+  */
+  record semanticVersion {
+    /* The major version number.  For version ``2.0.1``, this would be
+       ``2``. */
+    param major: int;
+
+    /* The minor version number.  For version ``2.0.1``, this would be
+       ``0``. */
+    param minor: int;
+
+    /* The update version number.  For version ``2.0.1``, this would
+       be ``1``. */
+    param update: int;
+
+    pragma "no doc"
+    proc writeThis(s) throws {
+      s.write(this:string);
+    }
+  }
+
+  // cast from semanticVersion to string
+  pragma "no doc"
+  proc _cast(type t: string, x: semanticVersion(?)) param : string {
+    return "version " + x.major:string + "." + x.minor:string + "." +
+           x.update:string;
+  }
+
+
+  /*
+    This record represents a software version in a Git repository.  It
+    uses ``param`` values to represent its components in order to
+    support compile-time comparison of version numbers which in turn
+    permits code to specialize to specific versions of Chapel.  When
+    printed or converted to a string, it is represented as ``version
+    major.minor.update (commit)``.
+
+    Note that comparisons between two sourceVersion values with identical
+    semantic versions are not supported due to the challenges involved
+    in ordering commit values.  In addition, when a semanticVersion is
```

Comment:
I think of `==` and `!=` as comparisons but these are implemented. Maybe say "comparisons other than `==` and `!=`"

Reply:
Good point.  Does using the term "ordered comparisons" fix this for you?

Reply:
I think that would be good enough, yes